dll/shellext/cabview/folder.cpp

```diff
@@ -0,0 +1,837 @@
+                psr->uType = STRRET_CSTR;
+                if (data & FILE_ATTRIBUTE_READONLY) psr->cStr[i++] = 'R';
+                if (data & FILE_ATTRIBUTE_HIDDEN) psr->cStr[i++] = 'H';
+                if (data & FILE_ATTRIBUTE_SYSTEM) psr->cStr[i++] = 'S';
+                if (data & FILE_ATTRIBUTE_ARCHIVE) psr->cStr[i++] = 'A';
+                psr->cStr[i++] = '\0';
+            }
+            return S_OK;
+        }
+    }
+    return hr;
+}
+
+IFACEMETHODIMP CCabFolder::GetDetailsEx(PCUITEMID_CHILD pidl, const SHCOLUMNID *pscid, VARIANT *pv)
+{
+    if (!pscid || !pv)
+        return E_INVALIDARG;
+
+    CABITEM *item;
+    int col = MapSCIDToColumn(*pscid);
+    if (col >= 0)
+    {
+        return GetItemDetails(pidl, col, NULL, pv);
+    }
+    else if ((item = CABITEM::Validate(pidl)) == NULL)
+    {
+        return E_INVALIDARG;
+    }
+    else if (IsEqual(*pscid, FMTID_ShellDetails, PID_FINDDATA))
+    {
+        WIN32_FIND_DATA wfd;
+        ZeroMemory(&wfd, sizeof(wfd));
+        wfd.dwFileAttributes = item->GetFSAttributes();
+        wfd.nFileSizeLow = item->Size;
+        DosDateTimeToFileTime(item->Date, item->Time, &wfd.ftLastWriteTime);
+        lstrcpyn(wfd.cFileName, item->GetName(), MAX_PATH);
+        return InitVariantFromBuffer(&wfd, sizeof(wfd), pv);
+    }
+    return E_FAIL;
+}
+
+IFACEMETHODIMP CCabFolder::GetDetailsOf(PCUITEMID_CHILD pidl, UINT iColumn, SHELLDETAILS *psd)
+{
+    if (!psd || iColumn >= _countof(g_Columns))
+    {
+        return E_INVALIDARG;
+    }
+    else if (!pidl)
+    {
+        psd->fmt = g_Columns[iColumn].LvcFmt;
+        psd->cxChar = g_Columns[iColumn].LvcChars;
+        WCHAR buf[MAX_PATH];
+        if (LoadStringW(g_hInst, g_Columns[iColumn].TextId, buf, _countof(buf)))
+            return StrTo(buf, psd->str);
+        return E_FAIL;
+    }
+    return GetItemDetails(pidl, iColumn, psd, NULL);
+}
+
+int CCabFolder::MapSCIDToColumn(const SHCOLUMNID &scid)
+{
+    for (UINT i = 0; i < _countof(g_Columns); ++i)
+    {
+        if (g_Columns[i].pkg && IsEqual(scid, *g_Columns[i].pkg, g_Columns[i].pki))
+            return i;
+    }
+    return -1;
+}
+
+IFACEMETHODIMP CCabFolder::MapColumnToSCID(UINT column, SHCOLUMNID *pscid)
+{
+    if (column < _countof(g_Columns) && g_Columns[column].pkg)
+    {
+        pscid->fmtid = *g_Columns[column].pkg;
+        pscid->pid = g_Columns[column].pki;
+        return S_OK;
+    }
+    return E_FAIL;
+}
+
+IFACEMETHODIMP CCabFolder::EnumObjects(HWND hwndOwner, DWORD dwFlags, LPENUMIDLIST *ppEnumIDList)
+{
+    CEnumIDList *p = CEnumIDList::CreateInstance();
+    *ppEnumIDList = static_cast<LPENUMIDLIST>(p);
+    return p ? p->Fill(m_CurDir, hwndOwner, dwFlags) : E_OUTOFMEMORY;
+}
+
+IFACEMETHODIMP CCabFolder::BindToObject(PCUIDLIST_RELATIVE pidl, LPBC pbcReserved, REFIID riid, LPVOID *ppvOut)
+{
+    UNIMPLEMENTED;
+    return E_NOTIMPL;
+}
+
+HRESULT CCabFolder::CompareID(LPARAM lParam, PCUITEMID_CHILD pidl1, PCUITEMID_CHILD pidl2)
```

Comment:
Should this be able to handle the case where `pidl1` or `pidl2` is `NULL` ?

Reply:
This is an internal function to compare a single level PIDL, `CompareIDs` is the one that validates its input (those parameters are not marked as optional though).